Add 'onFieldBeforePrepare' and 'onFieldAfterPrepare' events

williamsandy opened the pull request:
Added Events 'onFieldBeforeRender' and 'onFieldAfterRender' to
FieldsHelper render method to give extensions that are not necessary
components or templates the ability to modify the output of the rendered
custom fields.

Why would we want to do this?
Well, there are a few reasons why we want to do this, and one real life
example would be to do with inline editing. The two events would make it
easier to make the rendered field editable in the front end when the
user logs in and wants to edit their article or item.

Why do we need two events?

Well, the onFieldBeforeRender will allow modification directly to the
value of the custom field before it is rendered with the layout. And in
some cases, this will be needed.

The onFieldAfterRender event is there for the case when we can wait
until the custom field is fully rendered before doing any modification.

Both events take the context string and the display data so the plugin
can know what component is rendering the extra field, and from the data
object know the field type.
We also supply the layout file name reference so we can quickly know
what type of rendering is occurring. This will be set to 'mustache' when
we are using mustache syntax tags

Note: No manipulation of the display object is allowed in the
onFieldAfterRender event as we are only using it as a reference at this
stage.

williamsandy committed:
Add 'onFieldBeforeRender' and 'onFieldAfterRender' events to the field renderer for Custom Fields

Added Events 'onFieldBeforeRender' and 'onFieldAfterRender' to
FieldsHelper render method to give extensions that are not necessary
components or templates the ability to modify the output of the rendered
custom fields.

Why would we want to do this?
Well, there are a few reasons why we want to do this, and one real life
example would be to do with inline editing. The two events would make it
easier to make the rendered field editable in the front end when the
user logs in and wants to edit their article or item.

Why do we need two events?

Well, the onFieldBeforeRender will allow modification directly to the
value of the custom field before it is rendered with the layout. And in
some cases, this will be needed.

The onFieldAfterRender event is there for the case when we can wait
until the custom field is fully rendered before doing any modification.

Both events take the context string and the display data so the plugin
can know what component is rendering the extra field, and from the data
object know the field type.
We also supply the layout file name reference so we can quickly know
what type of rendering is occurring. This will be set to 'mustache' when
we are using mustache syntax tags

Note: No manipulation of the display object is allowed in the
onFieldAfterRender event as we are only using it as a reference at this
stage.

williamsandy commented:
Hi Allon Moritz,

I just added a pull requests for the changes I talked about at the JAB16 conference. Basically I added two events: onFieldBeforeRender and onFieldAfterRender. This will help extensions such as inline editing integrate with custom fields.

Andrew(Ark Extenstions)

laoneo commented on administrator/components/com_fields/helpers/fields.php:
```diff
@@ -161,6 +161,9 @@ public static function getFields ($context, $item = null, $prepareValue = false)
 				$field->rawvalue = $field->value;
+				
+				//Store item id on field instance for later use			
+				$field->item_id = $item->id;
```
This needs to be removed, as the fields are cached. I would pass the item as argument on the event directly.

williamsandy replied:
I don't see the difference in setting the rawvalue /value and setting the article/item id. Yes the field data are cached but they will always be set per article and the value be specific to the current article that is being rendered.

I have tested this with a test plugin for the editor field type  and a several of other field types and it seems to be working for me.

I would have passed the item into the event directly but in the render method we do not have access to the item when rendering the specific field unless we set the item id in the request which I thought was a bit overkill. Although I am  happy to do that if you think it is better doing it that way.

laoneo replied:
True, overlooked that.

laoneo replied:
What about removing the item id assignment here and adding the whole item as parameter to the trigger. Something like 

```
$dispatcher->trigger('onFieldBeforePrepare',array($context, $item, &$field));
```

and then on also

```
$dispatcher->trigger('onFieldAfterPrepare',array($context, $item, $field, &$value));
```

williamsandy replied:
I guess we can do that since the triggers to those events are now only in the getFields method.

laoneo commented on administrator/components/com_fields/helpers/fields.php:
```diff
@@ -170,8 +173,23 @@ public static function getFields ($context, $item = null, $prepareValue = false)
+							 */
+                            $dispatcher = JEventDispatcher::getInstance();
+							if($field->value)
+								$dispatcher->trigger('onFieldBeforeRender',array($context,'mustache',&$field));
```
Is the form of rendering really needed? The triggers should be independent of the form of rendering. Or do I miss here something?

williamsandy replied:
I put the triggers in the render method because at that point we have all the information we need. unfortunately for developers like me who have an extension that is a plugin and not a component or a template we will not be able to change the layout of the rendered custom field with a layout override. And in most cases even if  we  could get to the layout it will be too late by then as the information  we would need will not be present at that stage Or the render data will by in a format to late for us to change. 

So a real life example will be inline editing when we need to make the values of the custom fields clickable editable regions when the user is logged into the frontend of they website and want to update a custom field.

However, if you have an better ideas I am open to another way of doing it.

laoneo replied:
WHat I wanted to ask, is the $layoutFile/'mustache' parameter really needed? Would be nice to avoid it.

williamsandy replied:
The renderer context parameter was added as I thought it would be a nice to give the developer the option to respond differently if the rendering is done via Mustache i.e. by the end user or via  the component or template override. This is not essential, but I thought it could be useful in some cases. 

If you want that parameter removed, I am happy to remove that parameter.

laoneo replied:
Makes the event more understandable. And why not moving them to the render function? That would be the most obvious place IMO.

williamsandy replied:
I believe the events in the render function are not sufficient enough. Because the end user can override this and define a template in the output parameter of the desired custom field using Mustache syntax and thus it will only be the Mustache engine rendering the layout. So that is why I had to place the events there also.

williamsandy commented:
I think I may have misunderstood you sorry. Do you mean putting the event in the render function of the Mustache_Engine?

laoneo commented:
No, all events should be triggered in the getFields function. With the prepare instead render name.

williamsandy commented:
Okay so you want the events in the getFields to be called 'onFieldBeforePrepare' and 'onFieldAfterPrepare' and drop the mustache/layout parameter in those events.

And then have as before the events  onFieldBeforeRender and onFieldAfterRender in the render function

laoneo commented:
> Okay so you want the events in the getFields to be called 'onFieldBeforePrepare' and 'onFieldAfterPrepare' and drop the mustache/layout parameter in those events.

Yes

> And then have as before the events onFieldBeforeRender and onFieldAfterRender in the render function

No remove them completely.

williamsandy commented:
Okay I will make the change and ensure those events are called each time when the prepareValue flag is passed into the function. I will test it.

williamsandy committed:
Update fields.php

williamsandy committed:
Update fields.php

williamsandy committed:
Updates after discussion

Renamed events 'onFieldBeforeRender' and 'onFieldAfterRender'  to 'onFieldBeforePrepare' and 'onFieldAfterPrepare' respectively. Events are now always called in the getFields function whenever the PrepareValue flag is set to true. As a result removed onFieldBeforeRender and onFieldAfterRender event triggers from render function.

williamsandy commented:
I have tested and updated the code.

laoneo commented:
Looks much nicer. Guess the intendation is on some parts not correct.

williamsandy committed:
Pass Item instance directly into events: onFieldBeforePrepare and onFieldAfterPrepare

williamsandy committed:
Update fields.php

williamsandy committed:
Update fields.php

williamsandy commented:
I have updated and commit changes including removing all Travis errors from that file.

williamsandy commented:
Thanks for merging the change. Is it possible to add the change to DPFields as well as this is more or less a one to one copy. Do you want me to make a pull request for that project?

laoneo commented:
Yes, make there a PR as well and reference this one.
On Jun 1, 2016 1:20 PM, "Andrew" notifications@github.com wrote:

> Thanks for merging the change. Is it possible to add the change to
> DPFields as well as this is more or less a one to one copy. Do you want me
> to make a pull request for that project?
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> https://github.com/joomla-projects/custom-fields/pull/90#issuecomment-222964070,
> or mute the thread
> https://github.com/notifications/unsubscribe/AAPUwIu1ptSXujPxQUD3VOmmEcXe38Njks5qHWsRgaJpZM4IndHr
> .
